daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -726,11 +731,14 @@ object Speedy {
   }
 
   /** The function-closure and arguments have been evaluated. Now execute the body. */
-  final case class KFun(closure: PClosure, actuals: util.ArrayList[SValue])
+  final case class KFun(closure: PClosure, actuals: util.ArrayList[SValue], envSize: Int)
       extends Kont
       with SomeArrayEquals {
     def execute(v: SValue, machine: Machine) = {
       actuals.add(v)
+      // we only need to call restoreEnvStack (not the full restoreEnv), because frame/actuals
+      // are changed anyway to evaluate the function body.
+      machine.restoreEnvStack(envSize)
       // Set frame/actuals to allow access to the function arguments and closure free-varables.
       machine.frame = closure.frame
       machine.actuals = actuals
```

Comment:
We're literally doing here what `restoreEnv` does, aren't we?
```suggestion
      machine.restoreEnv(closure.frame, actuals, envSize)
```

Reply:
I hadn't considered using `restoreEnv` here, because in this case there was no corresponding save. But this is much much nicer. Thank you.